ogre2/src/Ogre2SegmentationCamera.cc

```diff
@@ -0,0 +1,647 @@
+
+  // random color
+  int r = distribution(this->generator);
+  int g = distribution(this->generator);
+  int b = distribution(this->generator);
+
+  math::Color color(r, g, b);
+
+  if (this->type == SegmentationType::Semantic)
+  {
+    // if the label is colored before return the color
+    // don't check fo taken colors in that case, all items
+    // with the same label will have the same color
+    if (this->coloredLabel.count(_label))
+      return color;
+  }
+
+  // loop recursivly till finding a unique color
+  if (this->IsTakenColor(color))
+      return this->LabelToColor(_label);
+
+  this->coloredLabel[_label] = true;
+  return color;
+}
+
+////////////////////////////////////////////////
+void SegmentationMaterialSwitcher::preRenderTargetUpdate(
+    const Ogre::RenderTargetEvent &/*_evt*/)
+{
+  this->datablockMap.clear();
+  auto itor = this->scene->OgreSceneManager()->getMovableObjectIterator(
+      Ogre::ItemFactory::FACTORY_TYPE_NAME);
+
+  while (itor.hasMoreElements())
+  {
+    Ogre::MovableObject *object = itor.peekNext();
+    Ogre::Item *item = static_cast<Ogre::Item *>(object);
+
+    // get visual from ogre item
+    Ogre::Any userAny = item->getUserObjectBindings().getUserAny();
+
+    if (!userAny.isEmpty() && userAny.getType() == typeid(unsigned int))
+    {
+      VisualPtr visual;
+      try
+      {
+        visual = this->scene->VisualById(Ogre::any_cast<unsigned int>(userAny));
+      }
+      catch(Ogre::Exception &e)
+      {
+        ignerr << "Ogre Error:" << e.getFullDescription() << "\n";
+      }
+      Ogre2VisualPtr ogreVisual = std::dynamic_pointer_cast<Ogre2Visual>(
+        visual);
+
+      // get class user data
+      Variant labelAny = ogreVisual->UserData(this->labelKey);
+
+      int label;
+      try
+      {
+        label = std::get<int>(labelAny);
+      }
+      catch(std::bad_variant_access &e)
+      {
+        // items with no class are considered background
+        label = this->backgroundLabel;
+      }
+
+      for (unsigned int i = 0; i < item->getNumSubItems(); i++)
+      {
+        // save subitems material
+        Ogre::SubItem *subItem = item->getSubItem(i);
+        Ogre::HlmsDatablock *datablock = subItem->getDatablock();
+        this->datablockMap[subItem] = datablock;
+
+        // Material Switching
+        if (this->type == SegmentationType::Semantic)
+        {
+          if (this->isColoredMap)
+          {
+            // semantic material(each pixel has item's color)
+            math::Color color = this->LabelToColor(label);
+            subItem->setCustomParameter(1, Ogre::Vector4(
+              color.R(), color.G(), color.B(), 1.0));
+          }
+          else
+          {
+            // labels ids material(each pixel has item's label)
+            float labelColor = label / 255.0;
+            subItem->setCustomParameter(1, Ogre::Vector4(
+              labelColor, labelColor, labelColor, 1.0));
```

Comment:
I'm concerned that this won't be very useful for users. Users want labels to be unique, and want labels to be something that is easy to work with (like an int). However, here, it looks like we are normalizing labels as floats between 0 and 1, right? If so, it would be difficult for users to compare pixels to see if they have the same label or not due to floating point roundoff error.

Reply:
It is in the range of [0-255], that normalization is just required to pass it to the function, but OGRE map it again to [0-255] in the buffer that the segmentation data is stored in